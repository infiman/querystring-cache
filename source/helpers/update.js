import { isPlainObject } from './isPlainObject'

export const update = (target, path, updater) => {
  if (process.env.NODE_ENV !== 'production') {
    if (!isPlainObject(target)) {
      throw new Error(
        `target param is not valid. Expecting: object! Received: ${Object.prototype.toString.call(
          target
        )}.`
      )
    }

    if (typeof path !== 'string') {
      throw new Error(
        `path param is not valid. Expecting: string! Received: ${Object.prototype.toString.call(
          path
        )}.`
      )
    }

    if (typeof updater !== 'function') {
      throw new Error(
        `updater param is not valid. Expecting: function! Received: ${Object.prototype.toString.call(
          updater
        )}.`
      )
    }
  }

  const hasValue = Object.prototype.hasOwnProperty.call(target, path)
  const oldValue = target[path]
  const updatedValue = updater(oldValue, path, { tail: true })

  if (!hasValue || oldValue !== updatedValue) {
    return Object.assign({}, target, { [path]: updatedValue })
  }

  return target
}

export const updateDeep = (target, path, updater, missingNodeResolver) => {
  if (process.env.NODE_ENV !== 'production') {
    if (!isPlainObject(target)) {
      throw new Error(
        `target param is not valid. Expecting: object! Received: ${Object.prototype.toString.call(
          target
        )}.`
      )
    }

    if (!Array.isArray(path)) {
      throw new Error(
        `path param is not valid. Expecting: array! Received: ${Object.prototype.toString.call(
          path
        )}.`
      )
    }

    if (typeof updater !== 'function') {
      throw new Error(
        `updater param is not valid. Expecting: function! Received: ${Object.prototype.toString.call(
          updater
        )}.`
      )
    }
  }

  let resolveMissingNode = missingNodeResolver || (() => ({}))
  let updated = Object.assign({}, target)
  let currentNode = updated
  let previousNode

  for (let i = 0, length = path.length; i < length; i++) {
    if (i === length - 1) {
      if (previousNode) {
        const updatedNode = update(currentNode, path[i], updater)

        if (currentNode === updatedNode) {
          return target
        }

        previousNode[path[i - 1]] = updatedNode
      } else {
        return update(target, path[i], updater)
      }
    } else {
      const oldOrMissingNode = currentNode[path[i]]
        ? currentNode[path[i]]
        : resolveMissingNode(path[i])
      const updatedNode = updater(oldOrMissingNode, path[i], { tail: false })

      currentNode[path[i]] = Object.assign({}, updatedNode)
      previousNode = currentNode
      currentNode = currentNode[path[i]]
    }
  }

  return updated
}
