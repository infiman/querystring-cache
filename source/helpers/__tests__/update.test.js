import { update, updateDeep } from '../update'

describe('update module', () => {
  describe('update', () => {
    test('empty invocation', () =>
      expect(() => update()).toThrowErrorMatchingSnapshot())

    test('one param', () =>
      expect(() => update({})).toThrowErrorMatchingSnapshot())

    test('two params', () =>
      expect(() => update({}, 'a')).toThrowErrorMatchingSnapshot())
  })

  describe('updateDeep', () => {
    test('empty invocation', () =>
      expect(() => updateDeep()).toThrowErrorMatchingSnapshot())

    test('one param', () =>
      expect(() => updateDeep({})).toThrowErrorMatchingSnapshot())

    test('two params', () =>
      expect(() => updateDeep({}, [])).toThrowErrorMatchingSnapshot())

    test('flat plain object', () =>
      expect(updateDeep({ a: 'c' }, ['a'], () => 'updated')).toEqual({
        a: 'updated'
      }))

    test('empty plain object', () =>
      expect(
        updateDeep({}, ['a', 'b', 'c'], (oldValue, key, { tail }) =>
          tail ? key : oldValue
        )
      ).toEqual({
        a: { b: { c: 'c' } }
      }))

    test('nested plain object', () =>
      expect(
        updateDeep(
          { a: { b: { c: 'c' } } },
          ['a', 'b', 'c'],
          (oldValue, _, { tail }) => (tail ? 'updated' : oldValue)
        )
      ).toEqual({ a: { b: { c: 'updated' } } }))

    test('no changes', () => {
      const target = { a: { b: { c: 'c' } } }

      expect(
        updateDeep(target, ['a', 'b', 'c'], (oldValue, _, { tail }) =>
          tail ? 'c' : oldValue
        )
      ).toBe(target)
    })

    test('immutable', () => {
      const target = { a: { b: { c: 'c' } } }

      expect(updateDeep(target, ['a', 'b', 'c'], () => 'updated')).not.toBe(
        target
      )
    })
  })
})
