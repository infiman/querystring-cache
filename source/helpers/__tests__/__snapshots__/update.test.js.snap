// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update module update empty invocation 1`] = `"target param is not valid. Expecting: object! Received: [object Undefined]."`;

exports[`update module update one param 1`] = `"path param is not valid. Expecting: string! Received: [object Undefined]."`;

exports[`update module update two params 1`] = `"updater param is not valid. Expecting: function! Received: [object Undefined]."`;

exports[`update module updateDeep empty invocation 1`] = `"target param is not valid. Expecting: object! Received: [object Undefined]."`;

exports[`update module updateDeep one param 1`] = `"path param is not valid. Expecting: array! Received: [object Undefined]."`;

exports[`update module updateDeep two params 1`] = `"updater param is not valid. Expecting: function! Received: [object Undefined]."`;
